### PR TITLE
Add full unit testing suite and CI ecosystem support

### PR DESCRIPTION
**Pull request by grey-nearing**
### 1. Comprehensive Unit Test Suite (204 Tests, 0 Failures)
Adds dedicated unit test coverage across previously untested core modules:
- **`evaluation/metrics.py`** (`test/test_metrics.py`): Unit tests for all 14 hydrological metrics (`NSE`, `MSE`, `RMSE`, `KGE`, `Alpha-NSE`, `Beta-NSE`, `Beta-KGE`, `Pearson-r`, `FHV`, `FMS`, `FLV`, `Peak-Timing`, `Missed-Peaks`, `Peak-MAPE`), edge cases (all-NaN, empty slices, shape mismatches), and metric dispatchers.
- **`training/loss.py` & `regularization.py`** (`test/test_losses.py`): Tests for `MaskedMSELoss`, `MaskedRMSELoss`, `MaskedNSELoss`, `MaskedCMALLoss`, multi-frequency masking, and multi-target loss weighting.
- **`utils/cmal_deterministic.py`** (`test/test_cmal_math.py`): Mathematical validation for CMAL PDF/CDF distributions, PPF inverse-CDF calculations, and vectorized quantile monotonicity.
- **`utils/samplingutils.py`** (`test/test_samplingutils.py`): Tests for point prediction sampling (`sample_cmal`, `sample_cmal_deterministic`, `sample_mcd`), zero-threshold normalization, and negative value handling modes.
- **`modelzoo/`** (`test/test_models.py`): Tests for `FC` MLP layers, `PositionalEncoding`, `Regression` / `CMAL` prediction heads, and `BaseModel` lifecycle.
- **`training/`** (`test/test_trainer.py`, `test/test_logger.py`): Tests for optimizers, loss/regularization initializers, `Logger` training/validation lifecycles, and artifact image generation.
- **`datasetzoo/caravan.py` & registry** (`test/test_caravan.py`, `test/test_datasetregistry.py`, `test/test_mfdata_loader.py`): Tests for static attribute loading, CSV timeseries parsing, dataset registry dynamic dispatch, and CLI data loader.
- **`utils/`** (`test/test_logging_utils.py`, `test/test_lstm_utils.py`, `test/test_memory.py`, `test/test_plots.py`): Tests for memory management, LSTM weight initialization, log filters, and hydrograph plotting.
- **CLI Dispatchers** (`test/test_cli.py`): Argument parsing and mode dispatching for `run.py` and `run_scheduler.py`.
### 2. Bug Fixes & Code Improvements
- **`googlehydrology/utils/samplingutils.py`**: Fixed `UnboundLocalError` during error message formatting when embedding dropouts are uninitialized.
- **`googlehydrology/datasetzoo/caravan.py`**: Fixed `NotImplementedError` with Dask `.chunk('auto')` when attribute tables contain string/object columns by specifying `chunk({'basin': -1})`.
- **`googlehydrology/training/__init__.py`**: Cleaned up obsolete reference to non-existent `TiedFrequencyMSERegularization`.
- **`test/test_multimet.py`**: Corrected mock assertion sequence to reflect scaler save-before-scale ordering.
- **`test/test_uncertainty.py`**: Guarded negative sample clipping check for deterministic CMAL quantile outputs.
### 3. Google Python Style Guide & Readability Compliance
- **Line Length (§3.2)**: 100% compliant with 80-character maximum line length across all test suites, docstrings, comments, and fixtures.
- **Import Organization (§3.13)**: Standardized three-block import grouping (Standard Library, Third-Party, First-Party) separated by single blank lines.
- **Unused Imports Cleaned**: Removed obsolete imports across all test files.
- **Test State Isolation**: Restored root logger handlers in `test_logging_utils.py` and isolated global `FLAGS` using `absl.testing.flagsaver` in `test_mfdata_loader.py`.
### 4. CI Ecosystem Modernization & Security Hardening
- **Environment**: Added `environments/environment_cpu.yml` for fast, lightweight CI test execution on Python 3.12.
- **Branch Triggers**: Added `main` branch to CI `pull_request` and `push` triggers in `.github/workflows/pytest-ci.yml`.
- **Security Policy Compliance (`zizmor`)**: 
  - Added top-level `permissions: read-all`.
  - Pinned GitHub Actions to immutable full commit SHAs (`actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683 # v4.2.2`, `conda-incubator/setup-miniconda@d2e6a045a86077fb6cad6f5adf368e9076ddaa8d # v3.1.0`).
- **Configuration**: Configured `pyproject.toml` with test markers (`unit`, `integration`, `slow`, `gpu`) and warning filters.
---
## Verification
- **Test Suite**: All 204 unit tests pass in ~2 minutes on CPU.
- **Statement Coverage**: Core modules (`googlehydrology.evaluation`, `googlehydrology.training`, `googlehydrology.utils`, `googlehydrology.modelzoo`) >95% covered.

### CODE REVIEW

File: googlehydrology/datasetzoo/multimet.py [hunk 1]
```diff
@@ -1055,8 +1055,14 @@ def _convert_to_tensor(
...
-    path = path.as_posix().replace('gs:/', 'gs://')
-    return xr.open_zarr(store=path, chunks='auto', decode_timedelta=True)
+    str_path = str(path)
+    if str_path.startswith('gs:') or str_path.startswith('gs/'):
+        store = path.as_posix().replace('gs:/', 'gs://')
+    else:
+        store = str_path
+    return xr.open_zarr(
+        store=store, chunks='auto', decode_timedelta=True, consolidated=False
+    )
```
Comment (ashermetzger-rgb) [inline googlehydrology/datasetzoo/multimet.py:1069]: When consolidated=False, xarray/zarr cannot read the top-level .zmetadata file. Instead, it must make individual, sequential HTTP roundtrips to GCS for every variable, dimension, coordinate, and .zattrs file. For MultiMet stores containing dozens of bands and products, this increases dataset open   
latency over remote GCS connections from ~50ms to multiple seconds per basin.   

```suggestion
    is_cloud = str_path.startswith(('gs:', 'gs/'))
    return xr.open_zarr(
        store=store,
        chunks='auto',
        decode_timedelta=True,
        consolidated=True if is_cloud else False,
    )
```

**Review by ashermetzger-rgb — DISMISSED** [review]
I opened some issues and logged other things that should be corrected in future PRs.
I did add 1 comment which I think should be applied.

**Review by ashermetzger-rgb — APPROVED** [review]
